src/rhythme/non-python/config.txt
# comments start with a hash
; ... or with a semicolon
; empty lines are fine as well

# Write n=label to select hardware channel number n (starting from 1)
# and attach a label to it. This will show up in the FieldTrip buffer
# and the GDF file.
# Currently, there is no check for double inclusion of channels.
# That is, if you define 1=FOO and 1=BAR, the first hardware channel
# will show up twice, with different names.
# Channels listed before the first [save] or [stream] line will be added
# to both saving and streaming selections. Write [select] to enable
# selection for both saving and streaming.
[select]
1=A1
2=A2
3=A3
4=A4
5=A5
6=A6
7=A7
8=A8
9=A9
10=A10
11=A11
12=A12
13=A13
14=A14
15=A15
16=A16
17=A17
18=A18
19=A19
20=A20
21=A21
22=A22
23=A23
24=A24
25=A25
26=A26
27=A27
28=A28
29=A29
30=A30
31=A31
32=A32
33=B1
34=B2
35=B3
36=B4
37=B5
38=B6
39=B7
40=B8
41=B9
42=B10
43=B11
44=B12
45=B13
46=B14
47=B15
48=B16
49=B17
50=B18
51=B19
52=B20
53=B21
54=B22
55=B23
56=B24
57=B25
58=B26
59=B27
60=B28
61=B29
62=B30
63=B31
64=B32
65=C1
66=C2
67=C3
68=C4
69=C5
70=C6
71=C7
72=C8
73=C9
74=C10
75=C11
76=C12
77=C13
78=C14
79=C15
80=C16
81=C17
82=C18
83=C19
84=C20
85=C21
86=C22
87=C23
88=C24
89=C25
90=C26
91=C27
92=C28
93=C29
94=C30
95=C31
96=C32
97=D1
98=D2
99=D3
100=D4
101=D5
102=D6
103=D7
104=D8
105=D9
106=D10
107=D11
108=D12
109=D13
110=D14
111=D15
112=D16
113=D17
114=D18
115=D19
116=D20
117=D21
118=D22
119=D23
120=D24
121=D25
122=D26
123=D27
124=D28
125=D29
126=D30
127=D31
128=D32
129=A1
130=A2
131=A3
132=A4
133=A5
134=A6
135=A7
136=A8
137=A9
138=A10
139=A11
140=A12
141=A13
142=A14
143=A15
144=A16
145=A17
146=A18
147=A19
148=A20
149=A21
150=A22
151=A23
152=A24
153=A25
154=A26
155=A27
156=A28
157=A29
158=A30
159=A31
160=A32
161=B1
162=B2
163=B3
164=B4
165=B5
166=B6
167=B7
168=B8
169=B9
170=B10
171=B11
172=B12
173=B13
174=B14
175=B15
176=B16
177=B17
178=B18
179=B19
180=B20
181=B21
182=B22
183=B23
184=B24
185=B25
186=B26
187=B27
188=B28
189=B29
190=B30
191=B31
192=B32
193=C1
194=C2
195=C3
196=C4
197=C5
198=C6
199=C7
200=C8
201=C9
202=C10
203=C11
204=C12
205=C13
206=C14
207=C15
208=C16
209=C17
210=C18
211=C19
212=C20
213=C21
214=C22
215=C23
216=C24
217=C25
218=C26
219=C27
220=C28
221=C29
222=C30
223=C31
224=C32
225=D1
226=D2
227=D3
228=D4
229=D5
230=D6
231=D7
232=D8
233=D9
234=D10
235=D11
236=D12
237=D13
238=D14
239=D15
240=D16
241=D17
242=D18
243=D19
244=D20
245=D21
246=D22
247=D23
248=D24
249=D25
250=D26
251=D27
252=D28
253=D29
254=D30
255=D31
256=D32


# Now we add some more channel for saving only
[save]


# And some channels we only want to stream
[stream]


# Please take note that the channels will be written out in the order specified here,
# that is, they will not be sorted according to the hardware channel number!

# Write "downsample X" with a positive number X to set the downsampling
# factor for streaming. The GDF file will receive full-rate data.
downsample 1

# Write "bandwidth X" with a positive real number X to set the cutoff frequency
# of a Butterworth low-pass filter.
# bandwidth 50

# Write "bworder N" to set the order of the lowpass Butterworth filter for downsampling
# bworder 4

# Refresh period (in seconds) for inserting extra events.
# Battery events are sent out in a fixed interval (set 0 to disable).
# Status events are sent out when the status changes, OR when the specified
# time has elapsed after the last event was sent out.
statusrefresh 4
batteryrefresh 20

# Triggers will be written as FieldTrip buffer events with type="TRIGGER" and a
# value corresponding to the 16-bit trigger signal.
# With the keyword splittrigger, you can opt to split the 16-bit trigger signal
# into two 8-bit signals, and give names to the events that are sent out for the
# low and high byte of the original signal, respectively.
# splittrigger stimulus response

src/rhythme/Configuration.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
10/16/2020

@author: jhyon

VERSION 1.0.0
"""
import numpy as np

try:
    from .Client import RHYTHME
except ImportError:
    from Client import RHYTHME
# import multiprocessing.pool
import toml
import argparse
import sys


# parser = argparse.ArgumentParser()
# parser.add_argument("--cf", help="Enter the path to the configuration toml file", required=True)
# args = parser.parse_args()

def run_rhythme(config_filepath, offline_filepath=None):
    # print('1')
    with open(str(config_filepath)) as fileObj:
        content = fileObj.read()
        all_config = toml.loads(content)

    # print(str(path))
    # insert path to directory where JH_offline_client.py is located
    # path = '/Users/justinhyon/Documents/GitHub/teamflow/src'
    # sys.path.insert(1, path)
    # print("all_config", all_config)
    # general configurations
    general_config = all_config['general_config']
    option = general_config['option']
    savepath = general_config['savepath']
    print_output = general_config['print_output']

    blocksize_sec = general_config['blocksize_sec']  # number of seconds per segment
    units = general_config['units']  # v for volts, mv for millivolts, uv for microvolts, Mv for megavolts
    nchansparticipant = general_config['nchansparticipant']  # number of channels per particiapnt
    numparticipants = general_config['numparticipants']
    removefirstsample = general_config['removefirstsample']
    resample_freq = general_config['resample_freq']

    # plotting configurations
    plotpref = general_config[
        'plotpref']  # use either 'participant'(teamflow score plot), 'experiment'(detailed plot), 'both', or 'none' to
    # toggle showing the plots
    ex_windowsize = general_config['ex_windowsize']  # size of the figures
    sub_windowsize = general_config['sub_windowsize']  # size of the figures
    ex_plot_dims = general_config[
        'ex_plot_dims']  # The dimensions of the plot grid to be used for the experimenter plot. Be sure to count total
    # number of required plots, or desired plots will not be rendered
    sub_plot_dims = general_config[
        'sub_plot_dims']  # The dimensions of the plot grid to be used for the subject plot. Be sure to count total
    # number of required plots, or desired plots will not be rendered
    saving = general_config[
        'saving']  # True or False. toggle saving the plots. Must create a folder called "TF_figures" in path directory
    # (line 13) before running. there is currently a bug that prevents the team flow scores plot from saving

    # channel configurations
    badchans = general_config['badchans']
    TrigChan = general_config[
        'TrigChan']  # python begins numbering from 0, so this is the channel number of the stim channel - 1 (not the name)
    # can be 'none' if there is no trigger channel, and a trigger channel of all 0s will be added to the end of the data
    # use option 'create' if events are being sent from FieldTrip Buffer, but there is no Trigger Channel in the data (ie.
    # realtime mode). This will create a stim channel with the events, and adds it as the last channel.

    channelnames = all_config['channelnames_dict']['channelnames']

    # configurations for realtime mode (ignored in offline mode)
    start_zero = general_config[
        'start_zero']  # whether the experiment must start from the first sample (empty buffer), or False for the current
    #  sample in the buffer
    exp_name = general_config['exp_name']
    dataport = general_config['dataport']  # use 3113 for eeg machine
    ignore_overflow = general_config['ignore_overflow']  # throw out samples past the desired block size
    n_skipped_segs = general_config[
        'n_skipped_segs']  # max number of segments to skip in between 2 processed segments due
    # to runtime
    wait_segs = general_config[
        'wait_segs']  # How many segments worth of time (blocksize_sec) to wait for new data before quitting the pipeline. does
    # not apply to the initial wait for first segment to arrive
    delay = general_config['delay']  # how long to wait before checking again for new samples in the buffer

    # configurations for offline mode (ignored in realtime mode)
    if offline_filepath:
        filepath = offline_filepath

    else:
        filepath = general_config['filepath']  # path to the  bdf data file
    fsample = general_config['fsample']  # samples per second

    function_dict = all_config['function_dict']

    num_plvsimevents = general_config['num_plvsimevents']  # number of simulated events for each PLV segment
    channelofint_plv = general_config['channelofint_plv']
    frqofint = general_config['frqofint']

    # function_dict['psd2'] = {
    #     'values_band1': [],
    #     'channelofint_band1': list(range(129, 161)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band1': 30,
    #     'foihigh_band1': 119,
    #     'plotwv_band1': 'experiment',
    #     'plotidx_band1': 'experiment',
    #
    #     'values_band2': [],
    #     'channelofint_band2': list(range(161, 193)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band2': 12,
    #     'foihigh_band2': 29,
    #     'plotwv_band2': 'experiment',
    #     'plotidx_band2': 'experiment',
    #
    #     'values_band3': [],
    #     'channelofint_band3': list(range(193, 225)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band3': 7,
    #     'foihigh_band3': 11,
    #     'plotwv_band3': 'experiment',
    #     'plotidx_band3': 'experiment',
    #
    #     'values_band4': [],
    #     'channelofint_band4': list(range(225, 256)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band4': 7,
    #     'foihigh_band4': 11,
    #     'plotwv_band4': 'experiment',
    #     'plotidx_band4': 'experiment',
    # }
    #
    # # function_dict['psd3'] = {
    # #     'values_band1': [],
    # #     'channelofint_band1': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band1': 30,
    # #     'foihigh_band1': 119,
    # #     'plotwv_band1': 'experiment',
    # #     'plotidx_band1': 'experiment',
    # #
    # #     'values_band2': [],
    # #     'channelofint_band2': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band2': 12,
    # #     'foihigh_band2': 29,
    # #     'plotwv_band2': 'experiment',
    # #     'plotidx_band2': 'experiment',
    # #
    # #     'values_band3': [],
    # #     'channelofint_band3': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band3': 7,
    # #     'foihigh_band3': 11,
    # #     'plotwv_band3': 'experiment',
    # #     'plotidx_band3': 'experiment',
    # # }
    # #
    # # function_dict['psd4'] = {
    # #     'values_band1': [],
    # #     'channelofint_band1': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band1': 30,
    # #     'foihigh_band1': 119,
    # #     'plotwv_band1': 'experiment',
    # #     'plotidx_band1': 'experiment',
    # #
    # #     'values_band2': [],
    # #     'channelofint_band2': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band2': 12,
    # #     'foihigh_band2': 29,
    # #     'plotwv_band2': 'experiment',
    # #     'plotidx_band2': 'experiment',
    # #
    # #     'values_band3': [],
    # #     'channelofint_band3': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band3': 7,
    # #     'foihigh_band3': 11,
    # #     'plotwv_band3': 'experiment',
    # #     'plotidx_band3': 'experiment',
    # # }
    #
    # function_dict['flow'] = {
    #             'values_Intra 1': [],
    #             'values_Intra 2': [],
    #             'values_Inter': []
    #         }
    #

    # # general configurations
    # option = 'realtime'
    # savepath = '../../non-python/misc/merge_test/rt'
    #
    # blocksize_sec = 5  # number of seconds per segment
    # units = 'uv'  # v for volts, mv for millivolts, uv for microvolts, Mv for megavolts
    # nchansparticipant = 128  # number of channels per particiapnt
    # numparticipants = 2
    # removefirstsample = True
    # resample_freq = 256
    #
    # # plotting configurations
    # plotpref = 'experiment'  # use either 'participant'(teamflow score plot), 'experiment'(detailed plot), 'both', or 'none' to
    #                             # toggle showing the plots
    # ex_windowsize = (24, 12)  # size of the figures
    # sub_windowsize = (36, 12)  # size of the figures
    # ex_plot_dims = (3, 4)  # The dimensions of the plot grid to be used for the experimenter plot. Be sure to count total
    #                         # number of required plots, or desired plots will not be rendered
    # sub_plot_dims = (2, 3)  # The dimensions of the plot grid to be used for the subject plot. Be sure to count total
    #                         # number of required plots, or desired plots will not be rendered
    # saving = False  # True or False. toggle saving the plots. Must create a folder called "TF_figures" in path directory
    #                 # (line 13) before running. there is currently a bug that prevents the team flow scores plot from saving
    #
    # # channel configurations
    # badchans = [ # must have list of badchans or empty list for each participant
    #     # ['A32', 'B2'], # participant 1
    #     [], # participant 2
    #     [],
    #     []
    # ]
    # TrigChan = 256  # python begins numbering from 0, so this is the channel number of the stim channel - 1 (not the name)
    # # can be 'none' if there is no trigger channel, and a trigger channel of all 0s will be added to the end of the data
    # # use option 'create' if events are being sent from FieldTrip Buffer, but there is no Trigger Channel in the data (ie.
    # # realtime mode). This will create a stim channel with the events, and adds it as the last channel.
    #
    # channelnames = [
    #
    # ]
    #
    # # THIS CODE GENERATES THE LIST OF CHANNEL NAMES USED IN OUR EXPERIMENT AND SHOULD NOT REMAIN IN THE FINAL CODE
    # # <<<<<<< Updated upstream
    # channelnum= 1
    # for p in range(2):
    #     thischannelnames = []
    #     for letter in ['A', 'B', 'C', 'D']:
    #         for num in range(1, 33):
    #             thischannelnames.append(letter + str(num))
    #
    #     channelnames.append(thischannelnames)
    # # =======
    # # channelnames = []
    # # >>>>>>> Stashed changes
    # # for p in range(2):
    # #     thischannelnames = []
    # #     for letter in ['A', 'B', 'C', 'D']:
    # #         for num in range(1, 33):
    # # <<<<<<< Updated upstream
    # #             channelnames.append(str(channelnum) + "=" + letter + str(num))
    # #             channelnum+=1
    # #     # channelnames.append(thischannelnames)
    # # print(channelnames)
    # # print ('\n'.join(channelnames))
    # # =======
    # #             thischannelnames.append(letter + str(num))
    # #     channelnames.append(thischannelnames)
    # # >>>>>>> Stashed changes
    #
    #
    # # configurations for realtime mode (ignored in offline mode)
    # start_zero = True  # whether the experiment must start from the first sample (empty buffer), or False for the current
    #                     #  sample in the buffer
    # exp_name = 'test_2'
    # dataport = 1972 # use 3113 for eeg machine
    # ignore_overflow = True # throw out samples past the desired block size
    # n_skipped_segs = 1  # max number of segments to skip in between 2 processed segments due to runtime
    # wait_segs = 2  # How many segments worth of time (blocksize_sec) to wait for new data before quitting the pipeline. does
    # # not apply to the initial wait for first segment to arrive
    # delay = .01  # how long to wait before checking again for new samples in the buffer
    #
    # # configurations for offline mode (ignored in realtime mode)
    # filepath = '/Users/jasonhyon/Desktop/JoinedData/170719_TRIAL2__EM_MD.set'  # path to the  bdf data file
    # fsample = 256  # samples per second
    #
    # function_dict = {}
    # # ----------------ERP configurations----------------- erp
    # #
    # function_dict['ERP1'] = {
    #     # subject 1
    #     'values_ERP': [],
    #     'channelofint': [1, 2, 3, 33, 34, 52, 65, 66, 87, 97, 98, 110, 111, 112],
    #     'epoeventval': 7,  # stim channel event value desired
    #     'pretrig': 1,  # seconds before event
    #     'posttrig': 1,  # seconds after event
    #     'bands': [[.110, .150], [.210, .250], [.310, .350]],
    #     'signs': ['-', '+', '-'], #  + for same sign of the average value in the corresponding band, - to reverse the sign, abs for absolute value
    #     'plotwv': 'experiment',
    #     'plotidx': 'experiment',
    #
    # }
    # function_dict['ERP2'] = {
    #     # subject 2
    #     'values_ERP': [],
    #     'channelofint': [129, 130, 131, 161, 162, 180, 193, 194, 215, 225, 226, 238, 239, 240],
    #     'epoeventval': 5,
    #     'pretrig': .3,
    #     'posttrig': .5,
    #     'bands': [[.110, .150], [.210, .250], [.310, .350]],
    #     'signs': ['-', '+', '-'],
    #     'plotwv': 'experiment',
    #     'plotidx': 'experiment'
    # }
    # # # function_dict['ERP3'] = {
    # # #     # subject 1
    # # #     'values_ERP': [],
    # # #     'channelofint': [1, 2, 3, 33, 34, 52, 65, 66, 87, 97, 98, 110, 111, 112],
    # # #     'epoeventval': 5,  # stim channel event value desired
    # # #     'pretrig': .3,  # seconds before event
    # # #     'posttrig': .5,  # seconds after event
    # # #     'bands': [[.110, .150], [.210, .250], [.310, .350]],
    # # #     'plotwv': 'experiment',
    # # #     'plotidx': 'experiment'
    # # # }
    # # #
    # # # function_dict['ERP4'] = {
    # # #     # subject 1
    # # #     'values_ERP': [],
    # # #     'channelofint': [1, 2, 3, 33, 34, 52, 65, 66, 87, 97, 98, 110, 111, 112],
    # # #     'epoeventval': 5,  # stim channel event value desired
    # # #     'pretrig': .3,  # seconds before event
    # # #     'posttrig': .5,  # seconds after event
    # # #     'bands': [[.110, .150], [.210, .250], [.310, .350]],
    # # #     'plotwv': 'experiment',
    # # #     'plotidx': 'experiment'
    # # # }
    # # # -----------------plv configurations (unused since MNE does everything)----------------- interbrain connectivity ibc
    # num_plvsimevents = 5  # number of simulated events for each PLV segment
    # channelofint_plv = 0
    # frqofint = 0
    # function_dict['plv'] = {
    #     'plot_option': 'experiment', # either tuple or one of 'experiment' or 'participant'. if tuple specified, intrabrain plv
    #                             # indexes will be plotted for the participants desired. if 'experiment' or 'participant' is
    #                             # specified, an average of all the intrabrain plvs is plotted
    #     # interbrain settings must come first
    #     'values_inter': [],
    #     'plot_matrix': 'experiment',
    #     'index_inter': 'experiment',
    #
    #     'values_intra1': [],
    #     'index_intra1': 'experiment',
    #
    #     'values_intra2': [],
    #     'index_intra2': 'experiment',
    #     #
    #     # 'values_intra3': [],
    #     # 'index_intra3': 'experiment',
    #     #
    #     # 'values_intra4': [],
    #     # 'index_intra4': 'experiment',
    #
    # }
    #
    # # # --------------psd configurations. T= temporal, F=frontal, A=occipital. 1 and 2 are the participant #s----------------
    # function_dict['psd1'] = {
    #     'values_band1': [],
    #     'channelofint_band1': list(range(32)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band1': 30,
    #     'foihigh_band1': 119,
    #     'plotwv_band1': 'experiment',
    #     'plotidx_band1': 'experiment',
    #
    #     'values_band2': [],
    #     'channelofint_band2': list(range(32, 64)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band2': 12,
    #     'foihigh_band2': 29,
    #     'plotwv_band2': 'experiment',
    #     'plotidx_band2': 'experiment',
    #
    #     'values_band3': [],
    #     'channelofint_band3': list(range(64, 97)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band3': 7,
    #     'foihigh_band3': 11,
    #     'plotwv_band3': 'experiment',
    #     'plotidx_band3': 'experiment',
    #
    #     'values_band4': [],
    #     'channelofint_band4': list(range(97, 129)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    #     'foilow_band4': 7,
    #     'foihigh_band4': 11,
    #     'plotwv_band4': 'experiment',
    #     'plotidx_band4': 'experiment',
    # }
    #
    # # function_dict['psd2'] = {
    # #     'values_band1': [],
    # #     'channelofint_band1': list(range(129, 161)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band1': 30,
    # #     'foihigh_band1': 119,
    # #     'plotwv_band1': 'experiment',
    # #     'plotidx_band1': 'experiment',
    # #
    # #     'values_band2': [],
    # #     'channelofint_band2': list(range(161, 193)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band2': 12,
    # #     'foihigh_band2': 29,
    # #     'plotwv_band2': 'experiment',
    # #     'plotidx_band2': 'experiment',
    # #
    # #     'values_band3': [],
    # #     'channelofint_band3': list(range(193, 225)),#[244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band3': 7,
    # #     'foihigh_band3': 11,
    # #     'plotwv_band3': 'experiment',
    # #     'plotidx_band3': 'experiment',
    # #
    # #     'values_band4': [],
    # #     'channelofint_band4': list(range(225, 256)),#[116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # #     'foilow_band4': 7,
    # #     'foihigh_band4': 11,
    # #     'plotwv_band4': 'experiment',
    # #     'plotidx_band4': 'experiment',
    # # }
    # #
    # # # function_dict['psd3'] = {
    # # #     'values_band1': [],
    # # #     'channelofint_band1': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band1': 30,
    # # #     'foihigh_band1': 119,
    # # #     'plotwv_band1': 'experiment',
    # # #     'plotidx_band1': 'experiment',
    # # #
    # # #     'values_band2': [],
    # # #     'channelofint_band2': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band2': 12,
    # # #     'foihigh_band2': 29,
    # # #     'plotwv_band2': 'experiment',
    # # #     'plotidx_band2': 'experiment',
    # # #
    # # #     'values_band3': [],
    # # #     'channelofint_band3': [116, 117, 119, 126],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band3': 7,
    # # #     'foihigh_band3': 11,
    # # #     'plotwv_band3': 'experiment',
    # # #     'plotidx_band3': 'experiment',
    # # # }
    # # #
    # # # function_dict['psd4'] = {
    # # #     'values_band1': [],
    # # #     'channelofint_band1': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band1': 30,
    # # #     'foihigh_band1': 119,
    # # #     'plotwv_band1': 'experiment',
    # # #     'plotidx_band1': 'experiment',
    # # #
    # # #     'values_band2': [],
    # # #     'channelofint_band2': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band2': 12,
    # # #     'foihigh_band2': 29,
    # # #     'plotwv_band2': 'experiment',
    # # #     'plotidx_band2': 'experiment',
    # # #
    # # #     'values_band3': [],
    # # #     'channelofint_band3': [244, 245, 247, 254],  # [117, 118, 120, 127] actual channels of interest subject 1
    # # #     'foilow_band3': 7,
    # # #     'foihigh_band3': 11,
    # # #     'plotwv_band3': 'experiment',
    # # #     'plotidx_band3': 'experiment',
    # # # }
    # #
    # # function_dict['flow'] = {
    # #             'values_Intra 1': [],
    # #             'values_Intra 2': [],
    # #             'values_Inter': []
    # #         }
    # #

    # if __name__ == '__main__':
    # if not print_output:
    #     sys.stdout = open(os.devnull, 'w')
    TF = RHYTHME(savepath, dataport, ex_windowsize, sub_windowsize, delay, plotpref, saving, TrigChan, )

    TF.master_control(filepath=filepath,
                      badchans=badchans,
                      fsample=fsample,
                      nchansparticipant=nchansparticipant,
                      trigchan=TrigChan,
                      blocksize_sec=blocksize_sec,
                      num_plvsimevents=num_plvsimevents,
                      option=option,
                      delay=delay,
                      units=units,
                      remfirstsample=removefirstsample,
                      exp_name=exp_name,
                      ignore_overflow=ignore_overflow,
                      n_skipped_segs=n_skipped_segs,
                      wait_segs=wait_segs,
                      function_dict=function_dict,
                      numparticipants=numparticipants,
                      ex_plot_matrix=np.zeros(ex_plot_dims),
                      sub_plot_matrix=np.zeros(sub_plot_dims),
                      channelnames=channelnames,
                      start_zero=start_zero,
                      resample_freq=resample_freq),
    sys.stdout = sys.__stdout__